### ISSUE TITLE
GRPO训练reward出现下降

### ISSUE THREAD

**Issue by Jingwei-Bao** [IDEA-Research/Rex-Omni#53]
您好，我在测试用GRPO训练一个简单的苹果识别任务，数据为数量为1～100的苹果图像（每个数量1000张，共10万组图文对，bbox使用绝对坐标）输出图中apple的bounding boxes：

<img width="640" height="640" alt="Image" src="https://github.com/user-attachments/assets/d4a2a91f-2e1d-4f43-8ebe-84a3149efbe2" />

但是reward（IoU）在训练后期出现奇怪的下降抖动现象：

<img width="1531" height="405" alt="Image" src="https://github.com/user-attachments/assets/464eda30-a303-4128-a70f-539c89621b14" />
我的训练参数如下，使用的基本是初始参数：
data:
  config_path: configs/grpo.py
  max_prompt_length: 4096
  max_response_length: 4096
  rollout_batch_size: 32
  val_batch_size: 16
  format_prompt: verl/configs/r1v_format.jinja
  override_chat_template: null
  shuffle: true
  seed: 1
  max_pixels: 1003520
  min_pixels: 12544
  filter_overlong_prompts: true

algorithm:
  adv_estimator: grpo
  disable_kl: false
  use_kl_loss: true
  kl_penalty: low_var_kl
  kl_coef: 1.0e-2

worker:
  actor:
    global_batch_size: 32
    micro_batch_size_per_device_for_update: 4
    micro_batch_size_per_device_for_experience: 16
    max_grad_norm: 1.0
    padding_free: true
    ulysses_sequence_parallel_size: 1
    model:
      model_path: /Rex-Omni/Rex-Omni    
      enable_gradient_checkpointing: true
      trust_remote_code: false
      freeze_vision_tower: false
    optim:
      lr: 1.0e-6
      weight_decay: 1.0e-2
      strategy: adamw  # {adamw, adamw_bf16}
      lr_warmup_ratio: 0.0
    fsdp:
      enable_full_shard: true
      enable_cpu_offload: false
      enable_rank0_init: true
    offload:
      offload_params: true  # true: more CPU memory; false: more GPU memory
      offload_optimizer: true  # true: more CPU memory; false: more GPU memory

  rollout:
    n: 8
    temperature: 1.0
    top_p: 0.99
    gpu_memory_utilization: 0.6
    enforce_eager: false
    enable_chunked_prefill: false
    tensor_parallel_size: 2
    max_num_batched_tokens: 16384
    limit_images: 1
    val_override_config:
      temperature: 0.5
      n: 1

  ref:
    fsdp:
      enable_full_shard: true
      enable_cpu_offload: true  # true: more CPU memory; false: more GPU memory
      enable_rank0_init: true
    offload:
      offload_params: false

  reward:
    reward_type: batch
    reward_function: verl/configs/reward_func.py:compute_score
    skip_special_tokens: false

trainer:
  total_epochs: 1
  max_steps: null
  project_name: easy_r1
  experiment_name: qwen2_5_7b_math_grpo
  logger: ["console", "wandb"]
  nnodes: 1
  n_gpus_per_node: 8
  val_freq: -1  # -1 to disable
  val_before_train: false
  val_only: false
  val_generations_to_log: 3
  save_freq: 64  # -1 to disable
  save_limit: -1  # -1 to disable
  save_checkpoint_path: /Rex-Omni/finetuning/work_dirs/grpo
  load_checkpoint_path: null

**Comment by Jingwei-Bao** [comment]
取1000次的ckpt去infer效果很好，准确率0.5-->0.9，IoU的reward很给力; 最后一次的ckpt效果较差，是因为RL后期训练不稳定吗，这方面有没有什么trick呀

**Comment by Jingwei-Bao** [comment]
> [@Jingwei-Bao](https://github.com/Jingwei-Bao) have you observed any improvement when using GRPO? Which checkpoint did you use for GRPO fine-tuning — the pretrained Rex-Omni model, or the one after phase 1 fine-tuning on your dataset?
> 
> [Rex-Omni/finetuning/scripts/grpo.sh](https://github.com/IDEA-Research/Rex-Omni/blob/8f936847793c7d6f5a9aff297c589e55a83ce96f/finetuning/scripts/grpo.sh#L26)
> 
> Line 26 in [8f93684](/IDEA-Research/Rex-Omni/commit/8f936847793c7d6f5a9aff297c589e55a83ce96f)
> 
>  MODEL_PATH="IDEA-Research/Rex-Omni"  # replace it with your local file path

I used my own dataset to train the Rex-Omni pretrained model using GRPO. I saw clear improvements about IoU and detection accuracy.

**Comment by John1231983** [comment]
So, you trained only Phase 2 (GRPO phase) without loading the SFT checkpoint trained on your dataset. Am I right?
Normally, the full training process includes two phases:

- Phase 1: SFT on your dataset, initialized from the pretrained Rex-Omni model.
- Phase 2: Load the SFT checkpoint from Phase 1 and continue training with GRPO.

From my understanding, you skipped Phase 1 and directly trained GRPO using the pretrained IDEA-Research/Rex-Omni model.

**Comment by John1231983** [comment]
@Mountchicken Thanks. Could you tell me which line in your code implement GRPO objective function

<img width="658" height="128" alt="Image" src="https://github.com/user-attachments/assets/ec00a198-e559-4ff4-9587-517050197699" />

**Comment by Mountchicken** [comment]
We are using this frame work https://github.com/hiyouga/EasyR1 for GRPO stage training. You may refer to this repo for more information.
